athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/common/PartitionUtil.java

```diff
@@ -0,0 +1,208 @@
+     * year={year}/month={month}
+     * Here, {year} and {month} are the partition key variable placeholders
+     */
+    private static final Pattern PARTITION_PATTERN = Pattern.compile(PARTITION_PATTERN_REGEX);
+
+    /**
+     * Match any alpha-num characters. Used to match VARCHAR type only partition keys
+     */
+    private static final String VARCHAR_OR_STRING_REGEX = "([a-zA-Z0-9_-\\s]+)";
+
+    private PartitionUtil()
+    {
+    }
+
+    /**
+     * Return a list of {@link AbstractMap.SimpleImmutableEntry<String, String>} instances
+     *
+     * @param table response of get table from AWS Glue
+     * @param partitionFolder      partition folder name
+     * @return List of {@link AbstractMap.SimpleImmutableEntry<String, String>} instances (column name, and value)
+     */
+    public static List<AbstractMap.SimpleImmutableEntry<String, String>> getPartitionColumnData(Table table, String partitionFolder)
+    {
+        Optional<String> optionalFolderRegex = getRegExExpression(table);
+        if (optionalFolderRegex.isPresent()) {
+            String folderRegex = optionalFolderRegex.get();
+            Pattern folderMatchPattern = Pattern.compile(folderRegex);
+            if (!folderMatchPattern.matcher(partitionFolder).matches()) {
+                return List.of();
+            }
+            return getStoragePartitions(table.getParameters().get(PARTITION_PATTERN_KEY),
+                    partitionFolder, folderRegex, table.getPartitionKeys());
+        }
+        return List.of();
+    }
+
+    /**
+     * Return a list of storage partition(column name, column type and value)
+     *
+     * @param partitionPattern Name of the bucket
+     * @param folderModel      partition folder name
+     * @param folderNameRegEx  folder name regular expression
+     * @param partitionColumns partition column name list
+     * @return List of storage partition(column name, column type and value)
+     */
+    protected static List<AbstractMap.SimpleImmutableEntry<String, String>> getStoragePartitions(String partitionPattern, String folderModel, String folderNameRegEx, List<Column> partitionColumns)
+    {
+        List<AbstractMap.SimpleImmutableEntry<String, String>> partitions = new ArrayList<>();
+        Matcher partitionPatternMatcher = PARTITION_PATTERN.matcher(partitionPattern);
+        Matcher partitionFolderMatcher = Pattern.compile(folderNameRegEx).matcher(folderModel);
+        while (partitionFolderMatcher.find()) {
+            for (int j = 1; j <= partitionFolderMatcher.groupCount() && partitionPatternMatcher.find(); j++) {
+                LOGGER.debug("Partition folder {} : {}", partitionPatternMatcher.group(1), partitionFolderMatcher.group(j));
+                Optional<AbstractMap.SimpleImmutableEntry<String, String>> optionalStoragePartition = produceStoragePartition(partitionColumns, partitionPatternMatcher.group(1), partitionFolderMatcher.group(j));
```

Comment:
It looks like `partitionColumns` is only being used when searching for whether or not a column name string exists in it, right?

So instead of just passing down `partitionColumns` and then searching through the list every time, you can do this instead:
- First after line 101, create a case insensitive `Set` like this:
```
var partitionColumnsSet = partitionColumns.stream()
    .map(c -> c.getName())
    .collect(Collectors.toCollection(() -> new java.util.TreeSet<String>(String.CASE_INSENSITIVE_ORDER)));
```
- Then on here on line 105-106, `produceStoragePartition` can removed and lines 105 and be simplified into:
```
if (partitionColumnsSet.contains(columnName)) {
    // If columnName is null the user will just get a NPE, which is fine since it shouldn't be null and your original code was throwing an exception anyway.
    partitions.add(new AbstractMap.SimpleImmutableEntry<>(column.getName(), columnValue));
} else {
    throw new IllegalArgumentException("Column '" + columnName + "' is not defined as partition key in Glue Table");
}
```

Additionally, for all of the situations (in all files) where we see:
`List<AbstractMap.SimpleImmutableEntry<String, String>>` (including where this type is nested under another List), have you considered whether or not these should just be `Map<String, String>` (they can even be a `TreeMap` if you need `String.CASE_INSENSITIVE_ORDER`).

  - Note: only do this if there's actually no problem and suits the situation naturally as it seems like it does here.

Reply:
Incorporated the changes.